File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/hub/HubDtos.kt
package app.pantopus.android.data.api.models.hub

import app.pantopus.android.data.api.models.common.JsonValue
import com.squareup.moshi.JsonClass

/** `GET /api/hub` — route `backend/routes/hub.js:24`. */
@JsonClass(generateAdapter = true)
data class HubResponse(
    val user: HubUser,
    val context: HubContext,
    val availability: HubAvailability,
    val homes: List<HubHomeSummary>,
    val businesses: List<HubBusinessSummary>,
    val setup: HubSetup,
    val statusItems: List<HubStatusItem>,
    val cards: HubCards,
    val jumpBackIn: List<HubJumpBackItem>,
    val activity: List<HubActivityItem>,
    val neighborDensity: HubNeighborDensity?,
)

@JsonClass(generateAdapter = true)
data class HubUser(
    val id: String,
    val name: String,
    val firstName: String?,
    val username: String,
    val avatarUrl: String?,
    val email: String,
)

@JsonClass(generateAdapter = true)
data class HubContext(
    val activeHomeId: String?,
    val activePersona: ActivePersona,
) {
    @JsonClass(generateAdapter = true)
    data class ActivePersona(val type: String)
}

@JsonClass(generateAdapter = true)
data class HubAvailability(
    val hasHome: Boolean,
    val hasBusiness: Boolean,
    val hasPayoutMethod: Boolean,
)

@JsonClass(generateAdapter = true)
data class HubHomeSummary(
    val id: String,
    val name: String,
    val addressShort: String,
    val city: String?,
    val state: String?,
    val latitude: Double?,
    val longitude: Double?,
    val isPrimary: Boolean,
    val roleBase: String,
)

@JsonClass(generateAdapter = true)
data class HubBusinessSummary(
    val id: String,
    val name: String,
    val username: String,
    val roleBase: String,
)

@JsonClass(generateAdapter = true)
data class HubSetup(
    val steps: List<Step>,
    val allDone: Boolean,
    val profileCompleteness: ProfileCompleteness,
) {
    @JsonClass(generateAdapter = true)
    data class Step(val key: String, val done: Boolean)

    @JsonClass(generateAdapter = true)
    data class ProfileCompleteness(
        val score: Double,
        val checks: Checks,
        val missingFields: List<String>,
    ) {
        @JsonClass(generateAdapter = true)
        data class Checks(
            val firstName: Boolean,
            val lastName: Boolean,
            val photo: Boolean,
            val bio: Boolean,
            val skills: Boolean,
        )
    }
}

@JsonClass(generateAdapter = true)
data class HubStatusItem(
    val id: String,
    val type: String,
    val pillar: String,
    val title: String,
    val subtitle: String?,
    val severity: String,
    val count: Int?,
    val dueAt: String?,
    val route: String,
    val entityRef: EntityRef?,
) {
    @JsonClass(generateAdapter = true)
    data class EntityRef(val kind: String, val id: String)
}

@JsonClass(generateAdapter = true)
data class HubCards(
    val personal: HubPersonalCard,
    val home: HubHomeCard?,
    val business: HubBusinessCard?,
)

@JsonClass(generateAdapter = true)
data class HubPersonalCard(
    val unreadChats: Int,
    val earnings: Double,
    val gigsNearby: Int,
    val rating: Double,
    val reviewCount: Int,
)

@JsonClass(generateAdapter = true)
data class HubHomeCard(
    val newMail: Int,
    val billsDue: List<HubBill>,
    val tasksDue: List<HubTask>,
    val memberCount: Int,
) {
    @JsonClass(generateAdapter = true)
    data class HubBill(
        val id: String,
        val name: String,
        val amount: Double,
        val dueAt: String,
    )

    @JsonClass(generateAdapter = true)
    data class HubTask(
        val id: String,
        val title: String,
        val dueAt: String,
    )
}

@JsonClass(generateAdapter = true)
data class HubBusinessCard(
    val newOrders: Int,
    val unreadThreads: Int,
    val pendingPayout: Double,
)

@JsonClass(generateAdapter = true)
data class HubJumpBackItem(
    val title: String,
    val route: String,
    val icon: String,
)

@JsonClass(generateAdapter = true)
data class HubActivityItem(
    val id: String,
    val pillar: String,
    val title: String,
    val at: String,
    val read: Boolean,
    val route: String,
)

@JsonClass(generateAdapter = true)
data class HubNeighborDensity(
    val count: Int,
    val radiusMiles: Double,
    val milestone: String?,
)

/**
 * `GET /api/hub/today` — provider-orchestrated, shape varies. Exposed as
 * untyped JSON until the provider bundle stabilises. Route:
 * `backend/routes/hub.js:596`.
 */
@JsonClass(generateAdapter = true)
data class HubTodayResponse(
    val today: JsonValue?,
    val error: String?,
)

/** `GET /api/hub/discovery` — route `backend/routes/hub.js:720`. */
@JsonClass(generateAdapter = true)
data class HubDiscoveryResponse(
    val items: List<DiscoveryItem>,
)

@JsonClass(generateAdapter = true)
data class DiscoveryItem(
    val id: String,
    val type: String,
    val title: String,
    val meta: String,
    val category: String,
    val avatarUrl: String?,
    val route: String,
)

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Hub/HubDTOs.swift
//
//  HubDTOs.swift
//  Pantopus
//
//  DTOs for the hub endpoints in `backend/routes/hub.js`. All field names
//  are preserved from the route responses (snake_case and camelCase both
//  appear verbatim; the decoder uses per-field coding keys, not
//  convertFromSnakeCase, so surprising case changes never happen).
//

import Foundation

// MARK: - Hub overview (GET /api/hub — backend/routes/hub.js:24)

/// `GET /api/hub` — the hub overview bundle.
public struct HubResponse: Decodable, Sendable, Hashable {
    public let user: HubUser
    public let context: HubContext
    public let availability: HubAvailability
    public let homes: [HubHomeSummary]
    public let businesses: [HubBusinessSummary]
    public let setup: HubSetup
    public let statusItems: [HubStatusItem]
    public let cards: HubCards
    public let jumpBackIn: [HubJumpBackItem]
    public let activity: [HubActivityItem]
    public let neighborDensity: HubNeighborDensity?

    public struct HubUser: Decodable, Sendable, Hashable, Identifiable {
        public let id: String
        public let name: String
        public let firstName: String?
        public let username: String
        public let avatarUrl: String?
        public let email: String
    }

    public struct HubContext: Decodable, Sendable, Hashable {
        public let activeHomeId: String?
        public let activePersona: ActivePersona

        public struct ActivePersona: Decodable, Sendable, Hashable {
            public let type: String
        }
    }

    public struct HubAvailability: Decodable, Sendable, Hashable {
        public let hasHome: Bool
        public let hasBusiness: Bool
        public let hasPayoutMethod: Bool
    }

    public struct HubHomeSummary: Decodable, Sendable, Hashable, Identifiable {
        public let id: String
        public let name: String
        public let addressShort: String
        public let city: String?
        public let state: String?
        public let latitude: Double?
        public let longitude: Double?
        public let isPrimary: Bool
        public let roleBase: String
    }

    public struct HubBusinessSummary: Decodable, Sendable, Hashable, Identifiable {
        public let id: String
        public let name: String
        public let username: String
        public let roleBase: String
    }

    public struct HubSetup: Decodable, Sendable, Hashable {
        public let steps: [HubSetupStep]
        public let allDone: Bool
        public let profileCompleteness: ProfileCompleteness

        public struct HubSetupStep: Decodable, Sendable, Hashable {
            public let key: String
            public let done: Bool
        }

        public struct ProfileCompleteness: Decodable, Sendable, Hashable {
            public let score: Double
            public let checks: Checks
            public let missingFields: [String]

            public struct Checks: Decodable, Sendable, Hashable {
                public let firstName: Bool
                public let lastName: Bool
                public let photo: Bool
                public let bio: Bool
                public let skills: Bool
            }
        }
    }

    public struct HubStatusItem: Decodable, Sendable, Hashable, Identifiable {
        public let id: String
        public let type: String
        public let pillar: String
        public let title: String
        public let subtitle: String?
        public let severity: String
        public let count: Int?
        public let dueAt: String?
        public let route: String
        public let entityRef: EntityRef?

        public struct EntityRef: Decodable, Sendable, Hashable {
            public let kind: String
            public let id: String
        }
    }

    public struct HubCards: Decodable, Sendable, Hashable {
        public let personal: HubPersonalCard
        public let home: HubHomeCard?
        public let business: HubBusinessCard?

        public struct HubPersonalCard: Decodable, Sendable, Hashable {
            public let unreadChats: Int
            public let earnings: Double
            public let gigsNearby: Int
            public let rating: Double
            public let reviewCount: Int
        }

        public struct HubHomeCard: Decodable, Sendable, Hashable {
            public let newMail: Int
            public let billsDue: [HubBill]
            public let tasksDue: [HubTask]
            public let memberCount: Int

            public struct HubBill: Decodable, Sendable, Hashable, Identifiable {
                public let id: String
                public let name: String
                public let amount: Double
                public let dueAt: String
            }

            public struct HubTask: Decodable, Sendable, Hashable, Identifiable {
                public let id: String
                public let title: String
                public let dueAt: String
            }
        }

        public struct HubBusinessCard: Decodable, Sendable, Hashable {
            public let newOrders: Int
            public let unreadThreads: Int
            public let pendingPayout: Double
        }
    }

    public struct HubJumpBackItem: Decodable, Sendable, Hashable {
        public let title: String
        public let route: String
        public let icon: String
    }

    public struct HubActivityItem: Decodable, Sendable, Hashable, Identifiable {
        public let id: String
        public let pillar: String
        public let title: String
        public let at: String
        public let read: Bool
        public let route: String
    }

    public struct HubNeighborDensity: Decodable, Sendable, Hashable {
        public let count: Int
        public let radiusMiles: Double
        public let milestone: String?
    }
}

// MARK: - Hub Today (GET /api/hub/today — backend/routes/hub.js:596)

/// `GET /api/hub/today` — provider-orchestrated response whose precise shape
/// varies by provider (weather, AQI, alerts, signals). We type only the
/// envelope and expose the inner payload as [`JSONValue`](x-source-tag://JSONValue).
public struct HubTodayResponse: Decodable, Sendable, Hashable {
    /// The untyped payload; `nil` when the service reports
    /// `error: CONTEXT_UNAVAILABLE`.
    public let today: JSONValue?
    /// Server-side error signal when context can't be assembled.
    public let error: String?
}

// MARK: - Hub Discovery (GET /api/hub/discovery — backend/routes/hub.js:720)

/// `GET /api/hub/discovery` response envelope.
public struct HubDiscoveryResponse: Decodable, Sendable, Hashable {
    public let items: [Item]

    public struct Item: Decodable, Sendable, Hashable, Identifiable {
        public let id: String
        public let type: String
        public let title: String
        public let meta: String
        public let category: String
        public let avatarUrl: String?
        public let route: String
    }
}
